backends/ebpf/runtime/pcap_util.h

```diff
@@ -0,0 +1,207 @@
+
+/* Interfaces are named by integers */
+typedef uint16_t iface_index;
+
+/* A network packet.
+   Contains packet content, timestamp and the interface where
+   the packet has been received/sent.
+ */
+typedef struct {
+    char *data;
+    struct pcap_pkthdr pcap_hdr;
+    iface_index ifindex;
+} pcap_pkt;
+
+struct pcap_list;
+struct pcap_list_array;
+typedef struct pcap_list pcap_list_t;
+typedef struct pcap_list_array pcap_list_array_t;
+
+/**
+ * @brief Retrieve packets from a pcap file.
+ * @details Retrieves a list of packets from a given pcap file.
+ * This function allocates a pcap_list handle and fills it
+ * iteratively with the packets contained in the capture. All data in the
+ * capture file is copied to the new list.
+ * Each packet is assigned the given interface index as meta-information.
+ * A list allocated by this function should subsequently be freed by
+ * delete_list().
+ *
+ * @param pcap_file_name The exact name of the pcap file.
+ * @param index Interface index of the file.
+ *
+ * @return A handle to the allocated list. NULL if the read operation
+ * was unsuccessful.
+ */
+pcap_list_t *read_pkts_from_pcap(const char *pcap_file_name, iface_index index);
+
+/**
+ * @brief Write a list of packets to a pcap file.
+ * @details Opens a "dead" pcap handle as Ethernet device and
+ * iteratively dumps the packets to the given filename.
+ *
+ * @param pcap_file_name Exact name of the file to create and write to.
+ * @param pkt_list List of packets to write.
+ *
+ * @return EXIT_FAILURE if operation fails.
+ */
+int write_pkts_to_pcap(const char *pcap_file_name, const pcap_list_t *pkt_list);
+
+/**
+ * @brief Merges a given list array into a single list.
+ * @details Expects a handle to an array list and allocates a single cohesive
+ * list from it. The array and its data structures are subsequently destroyed.
+ *
+ *
+ * @param array The array containing n lists.
+ * @return A single merged list contained all packets. NULL if the operation
+ * the fails.
+ */
+pcap_list_t *merge_pcap_lists(pcap_list_array_t *array);
+
+/**
+ * @brief Returns an array of lists from a single list split by interface.
+ * @details Splits the given list by the interface indicator in the
+ * packet structure. The input list is destroyed and for each interface a
+ * new list is allocated.
+ *
+ * @param pkt_list A list containing packets.
+ * @return Array containing the new lists.
+ */
+pcap_list_array_t *split_list_by_interface(pcap_list_t *pkt_list);
+
+/**
+ * @brief Appends a reference to a packet to a given list of packets.
+ * @details This function takes a pointer to a packet and appends it to the
+ * given list. If the list does not exist, it is allocated. For each new packet
+ * to append, the list is expanded by the size of the packet pointer.
+ *
+ * @param pkt_list List handle descriptor. Can be null
+ * @param pkt Pointer of the packet to append.
+ *
+ * @return Returns the updated handle of the list. This function causes an
```

Comment:
I had chosen exit instead of an error code since a remalloc error is fatal for the program anyway. It makes the code a bit simpler by avoiding the checks for errors for each append call. The pointer itself changes depending on the remalloc output iirc. I can make it a double pointer argument instead.

Reply:
exit is fine